langchain/load/serializable.py

```diff
@@ -0,0 +1,93 @@
+from abc import ABC, abstractmethod
+from typing import TypedDict, Literal, Any
+
+from pydantic import BaseModel
+
+
+class BaseSerialized(TypedDict):
+    lc: int
+    type: str
+    id: list[str]
+
+
+class SerializedConstructor(BaseSerialized):
+    type: Literal["constructor"]
+    kwargs: dict[str, Any]
+
+
+class SerializedSecret(BaseSerialized):
+    type: Literal["secret"]
+
+
+class SerializedNotImplemented(BaseSerialized):
+    type: Literal["not_implemented"]
+
+
+class Serializable(BaseModel, ABC):
+    @property
+    def lc_namespace(self) -> list[str]:
+        """
+        Return the namespace of the langchain object.
+        eg. ["langchain", "llms", "openai"]
+        """
+        return self.__class__.__module__.split(".")
+
+    @property
+    def lc_secrets(self) -> dict[str, str]:
+        """
+        Return a map of constructor argument names to secret ids.
+        eg. {"openai_api_key": "OPENAI_API_KEY"}
+        """
+        return dict()
+
+    lc_kwargs: dict[str, Any] = dict()
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.lc_kwargs = kwargs
```

Comment:
this means if the arguments are changed at all this will not be reflected right?

eg if I do:

```
llm = LLM(model_name="foo")
llm.model_name = "bar"
```
if i serialize it, it will still show foo

Reply:
my top of head suggestion would be for the class just to define exactly what attributes it should serialize

is somewhat equally brittle in that you have to make sure to keep that up to date as you add parameters tho

could also make this class NOT mutable?

Reply:
My preference would be to make all classes immutable if that is easy to do

Reply:
Otherwise the class marking attributes as serializable is exactly what lc_fields is for. Should we then keep that so that classes can opt into marking attributes as mutable ?